### PR TITLE
[Feature][Humming] Humming feature integration

### PR DESCRIPTION
**Pull request by jinzhen-lin**
This PR expands Humming quantization support and integrates a unified input processing kernel.

#### 1. Hadamard transforms for online quantization

Enable Hadamard transforms by setting `hadamard_block_size` in `VLLM_HUMMING_ONLINE_QUANT_CONFIG`.

#### 2. SM120/SM121 support

Add support for NVFP4, MXFP4, and MXFP8, including the non-public `float4e0m3` and `float8e3m4` data types, see https://github.com/vllm-project/humming/pull/41 .

#### 3. Additional input quantization modes

Support `static_per_tensor`, `static_tensor_dynamic_group`, and `dynamic_group_tensor`. The latter two primarily serve NVFP4 quantization.

#### 4. Checkpoint quantization and input schema fallback

Honor checkpoint input quantization settings and automatically fall back to a higher-precision input schema when the device does not support the configured activation quantization type.

To retain the original WNA16 behavior, set:

```bash
export VLLM_HUMMING_INPUT_QUANT_CONFIG='{"dtype":"bfloat16"}'
```

Input schemas explicitly specified through `VLLM_HUMMING_INPUT_QUANT_CONFIG` disable fallback by default. Unsupported configurations raise an error unless `"allow_fallback": true` is explicitly specified.

#### 5. Unified `process_input` kernel

Support fused combinations of optional activation function, Hadamard transforms, and input quantization across multiple layouts.

Custom activations support `unary`, `binary_split`, and `binary_interleaved` modes, with implementations supplied as CUDA C++ expressions.

| Path | Gemm Type | Fused operations | Padding handling |
|---|---|---|---|
| Dense | `dense` | Optional Hadamard transform and input quantization | — |
| MoE gate/up | `indexed` | Optional Hadamard transform and input quantization | — |
| MoE gate/up | `grouped_contiguous` | MoE permutation, optional Hadamard transform and input quantization | Skip invalid padding tokens in EP |
| MoE gate/up | `grouped_mask` | Optional Hadamard transform and input quantization | Skip invalid padding tokens |
| MoE down | `indexed` | Activation, optional Hadamard transform and input quantization | Skip invalid padding tokens in EP |
| MoE down | `grouped_contiguous` | Activation, optional Hadamard transform and input quantization | Skip invalid padding tokens in EP |
| MoE down | `grouped_mask` | Activation, optional Hadamard transform and input quantization | Skip invalid padding tokens |

#### 6. Additional scatter indices

Extend `moe_permute` and `moe_align_block_size` to return the `scatter_idx` required by `process_input`.

#### 7. Humming mxfp6 backend

Add `HummingMxFp6LinearKernel`.

#### 8. Humming utility refactoring

Refactor `humming_utils.py`, moving some validation and conversion logic into the Humming library.

#### 9. Type annotations

Add type annotations to `vllm/utils/humming.py`.

### CODE REVIEW

File: vllm/model_executor/kernels/linear/mxfp6/humming.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        x: torch.Tensor,
+        bias: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        return apply_humming_linear(
+            layer,
+            x,
+            layer_config=self.layer_config,
+            compute_config=self.compute_config,
+            locks=self.locks,
+        )
```
Comment (mgoin) [inline vllm/model_executor/kernels/linear/mxfp6/humming.py:84]: Looks like we ignore the bias arg here

File: vllm/model_executor/layers/quantization/utils/humming/schema.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+
+    if not key.symmetric:
+        raise ValueError("Humming input quantization must be symmetric")
+
+    quant_dtypes = {value: dtype for dtype, value in _HUMMING_TO_QUANT_DTYPE.items()}
+    quant_dtypes[FP4_DTYPE] = humming_dtypes.float4e2m1
+    quant_dtypes[FP8_DTYPE] = humming_dtypes.float8e4m3
+    scale_dtypes = {value: dtype for dtype, value in _HUMMING_TO_SCALE_DTYPE.items()}
+    if key.dtype not in quant_dtypes or key.scale.dtype not in scale_dtypes:
+        raise ValueError(f"Unsupported Humming input or scale dtype: {key}")
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/humming/schema.py:227]: Should we be referencing VLLM_HUMMING_INPUT_QUANT_CONFIG as a global override here? It seems from the callers in like `mxfp8/humming.py` that calls as `input_schema = quant_key_to_input_schema(kMxfp8Dynamic)` that it would always hardcode the activation schemas to that type.

File: vllm/model_executor/kernels/linear/mxfp6/humming.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            return False, "Humming only supported on SM75+"
+
+        return True, None
+
+    @classmethod
+    def can_implement(cls, config: MxFp6LinearLayerConfig) -> tuple[bool, str | None]:
+        if config.weight_quant_key not in (kMxfp6E2M3Static, kMxfp6E3M2Static):
+            return False, "only supports MXFP6 E2M3 or E3M2 weights"
+
+        return True, None
```
Comment (mgoin) [inline vllm/model_executor/kernels/linear/mxfp6/humming.py:44]: Isn't this missing a `quant_key_to_input_schema` check like mxfp4?

File: vllm/model_executor/kernels/linear/mxfp8/humming.py [hunk 4]
```diff
@@ -50,7 +52,10 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         }
 
         convert_linear_layer_to_humming_standard(layer=layer, name_map=name_map)
-        self.layer_config = prepare_humming_linear_layer_config(layer, quant_config)
+        input_schema = quant_key_to_input_schema(kMxfp8Dynamic)
```
Comment (mgoin) [inline vllm/model_executor/kernels/linear/mxfp8/humming.py:55]: We should not be hardcoding this activation type to mxfp8. For instance Hopper GPUs currently go through this backend for mxfp8 weights and run as w8a16. Since this activation type will now be mxfp8 activations too, I expect it would fail on Hopper due to lack of hardware support.

File: vllm/model_executor/kernels/linear/mxfp4/humming.py [hunk 5]
```diff
@@ -39,14 +36,10 @@ def is_supported(
...
-        if config.activation_quant_key is not None:
-            logger.warning_once(
-                "HummingMxFp4LinearKernel is a weight-only (A16) kernel; "
-                "the requested activation quantization (%s) is ignored.",
-                config.activation_quant_key,
-            )
+        try:
+            quant_key_to_input_schema(config.activation_quant_key)
+        except ValueError as error:
+            return False, str(error)
```
Comment (mgoin) [inline vllm/model_executor/kernels/linear/mxfp4/humming.py:42]: I think this isn't technically preserving the same behavior on main. Currently if `HummingMxFp4LinearKernel.can_implement` is called with `activation_quant_key=kMxfp4Dynamic` on hardware that doesn't support mxfp4, we just ignore the activation dtype and run as w4a16. I'm not sure if we always make sure to set `activation_quant_key=None` before calling, need to audit

File: vllm/model_executor/kernels/linear/mxfp6/humming.py (hunk 1 above)
Comment (jinzhen-lin) [reply vllm/model_executor/kernels/linear/mxfp6/humming.py:84]: `bias` is accessed in `apply_humming_linear` via `getattr(layer, "bias", None)`, so we don't need to pass `bias` here.

I added a comment to avoid confusion.

File: vllm/model_executor/kernels/linear/mxfp6/humming.py (hunk 3 above)
Comment (jinzhen-lin) [reply vllm/model_executor/kernels/linear/mxfp6/humming.py:44]: Yes, Thank you! Added now.

File: vllm/model_executor/kernels/linear/mxfp8/humming.py (hunk 4 above)
Comment (jinzhen-lin) [reply vllm/model_executor/kernels/linear/mxfp8/humming.py:55]: Humming now prioritizes the activation type specified in the checkpoint, and then internally converts it to an activation type supported by the current hardware. The activation type passed in here is the one specified in the checkpoint.

File: vllm/model_executor/kernels/linear/mxfp4/humming.py (hunk 5 above)
Comment (jinzhen-lin) [reply vllm/model_executor/kernels/linear/mxfp4/humming.py:42]: Humming now prioritizes the activation type specified in the checkpoint, and then internally converts it to an activation type supported by the current hardware.

The detailed logic is implemented here:

https://github.com/jinzhen-lin/vllm/blob/c3fe11aaf1adeba57e97508118ffe1eb8ef75285/vllm/model_executor/layers/quantization/utils/humming/schema.py#L262-L394

For example, for weights using MXFP4 weights and MXFP4 activations, we will prefer FP8 activations on Ada and Hopper.

Humming is not a weight-only backend. I think the default behavior should be to use the highest-performance precision available within the supported range. The previous default WNA16 mode was mainly due to Humming's input schema design and support not being fully implemented at the time.

If users want to preserve the previous default WNA16 behavior, they can use:

```bash
export VLLM_HUMMING_INPUT_QUANT_CONFIG='{"dtype":"bfloat16"}'
```

File: vllm/model_executor/kernels/linear/mxfp6/humming.py (hunk 1 above)
Comment (jinzhen-lin) [reply vllm/model_executor/kernels/linear/mxfp6/humming.py:84]: Updated: I found that the original implementation could cause an extra bias to be added when TP > 1. I've fixed it now.

**Review by mgoin — APPROVED** [review]
